core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java

```diff
@@ -18,843 +18,73 @@
-
-      /**
-       * Default buffer size.
-       */
-      private static final int DEFAULT_BUFFER_SIZE = 64 * 1024;
-
-      /**
-       * Whether or not the codec status has been checked. Ensures the default codec is not
-       * recreated.
-       */
-      private final AtomicBoolean checked = new AtomicBoolean(false);
-
-      private transient CompressionCodec codec = null;
-
-      @Override
-      public CompressionCodec getCodec() {
-        return codec;
-      }
-
-      @Override
-      public void initializeDefaultCodec() {
-        codec = initCodec(checked, DEFAULT_BUFFER_SIZE, codec);
-      }
-
-      /**
-       * Creates a new ZStandard codec.
-       */
-      @Override
-      protected CompressionCodec createNewCodec(final int bufferSize) {
-        return createNewCodec(CONF_ZSTD_CLASS, DEFAULT_CLAZZ, bufferSize, BUFFER_SIZE_OPT);
-      }
-
-      @Override
-      public OutputStream createCompressionStream(OutputStream downStream, Compressor compressor,
-          int downStreamBufferSize) throws IOException {
-        if (!isSupported()) {
-          throw new IOException(
-              "ZStandard codec class not specified. Did you forget to set property "
-                  + CONF_ZSTD_CLASS + "?");
-        }
-        return createFinishedOnFlushCompressionStream(downStream, compressor, downStreamBufferSize);
-      }
-
-      @Override
-      public InputStream createDecompressionStream(InputStream downStream,
-          Decompressor decompressor, int downStreamBufferSize) throws IOException {
-        if (!isSupported()) {
-          throw new IOException(
-              "ZStandard codec class not specified. Did you forget to set property "
-                  + CONF_ZSTD_CLASS + "?");
-        }
-        return createDecompressionStream(downStream, decompressor, downStreamBufferSize,
-            DEFAULT_BUFFER_SIZE, ZSTANDARD, codec);
-      }
-
-      @Override
-      public boolean isSupported() {
-        return codec != null;
-      }
-    };
-
-    /**
-     * Guava cache to have a limited factory pattern defined in the Algorithm enum.
-     */
-    private static LoadingCache<Entry<Algorithm,Integer>,CompressionCodec> codecCache =
-        CacheBuilder.newBuilder().maximumSize(25).build(new CacheLoader<>() {
-          @Override
-          public CompressionCodec load(Entry<Algorithm,Integer> key) {
-            return key.getKey().createNewCodec(key.getValue());
-          }
-        });
+  private static final Map<String,DefaultCompressionAlgorithm> CONFIGURED_ALGORITHMS =
+      new HashMap<>();
+
+  private static final ServiceLoader<CompressionAlgorithmConfiguration> COMPRESSION_ALGORITHMS =
+      ServiceLoader.load(CompressionAlgorithmConfiguration.class);
+
+  // All compression-related settings are required to be configured statically in the
+  // Configuration object.
+  protected static final Configuration conf;
```

Comment:
This isn't formatted, but the basic idea is there:

```suggestion
  protected static final Configuration conf = new Configuration();
  private static final ServiceLoader<CompressionAlgorithmConfiguration> COMPRESSION_ALGORITHMS =
      ServiceLoader.load(CompressionAlgorithmConfiguration.class);
  private static final Map<String,DefaultCompressionAlgorithm> CONFIGURED_ALGORITHMS =
      StreamSupport.stream(COMPRESSION_ALGORITHMS.spliterator(), false)
          .map(a -> new DefaultCompressionAlgorithm(a, conf))
          .collect(Collectors.toMap(algo -> algo.getName(), algo -> algo));
```

Reply:
Implemented this change in 362e7bc